### ISSUE TITLE
Proposed description of how to generate P4Info for `type` definitions in P4 programs

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4runtime#189]
OK, this took way too long to write, but hopefully it is clear enough for you to understand.  Corrections, questions, comments all welcome, especially on the weirdness that is Example 3.

https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files

**Comment by jafingerhut** [comment]
@hesingh See the link above for a description of a proposed way to determine what to write to part of a p4info file when there are `type` and/or `typedef` types in a P4_16 program.  Antonin has not reviewed it yet, so do not be surprised if he finds problems with it later.

I am pinging you here because they seem related to this PR: https://github.com/p4lang/p4c/pull/1736

Until these details are reviewed and approved, I would treat them as 'experimental', and would recommend that you not try to modify that PR yet based on anything written here so far.  Even if we get an approved version of this, it might make more sense for further changes to be made after the PR above is committed (I don't have an informed opinion on whether one PR or two is better here).

**Comment by hemant-avatar3** [comment]
@jafingerhut Sure, I didn't have any plans to change the PR code until Antonin asks for a change.   Having skimmed through the note, this is a larger churn than what the PR has and thus punting it makes sense right now.

A nitpick:

Where you say an `enum with an underlying type of bit<W>`, maybe just say, 'a SerEnum which is a serializable enum` .  Why not allow boolean in the mix for not just SerEnum but for other types.

**Comment by jafingerhut** [comment]
Antonin had the same question.  The word "serializable" currently appears nowhere in the P4_16 language specification.  The word "underlying" and "underlying type" appear multiple times in the spec in the section on `enum`s where the language feature is defined.  If someone wants to look up what an enum with an underlying type in the language spec is, they can find it.  If they try to look up SerEnum or serializable, they will find nothing.

That could be changed in the spec, of course, but going by what is there today, calling them anything else is an exercise in frustrating the reader looking for a definition.

`bool` could of course be added.  That would be a change to the current P4Runtime spec.  A small one, admittedly, but a change.

**Comment by hesingh** [comment]
Ok.  

Regarding

"Question: Should the P4Info file only include a type in the `new_types` section if the `type` is actually used in some other part of the P4Info file? Or should it include all types defined via a type declaration, even if it is not used elsewhere in the program?"

Hmm, during compilation, p4c removes any unused variables by p4c's frontend code.  p4runtime operates after p4 frontend runs.  Thus, I would think all unused types maybe removed as well.   One can test and check.

**Comment by jafingerhut** [comment]
Created PR proposing mentioning the word "serializable" at least once in the language spec, so people can search the spec for that word: https://github.com/p4lang/p4-spec

**Comment by jafingerhut** [comment]
Probably the correct answer to that question is that a type should only appear in the `new_types` section if it is actually used somewhere else in the program that has not been optimized away.

**Comment by antoninbas** [comment]
I don't think example 2 (https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#example-2) is correct. Same for this program: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/expected-p4info-files/p416-type-use3.v1model.p4info.txt

The P4Runtime spec says that for a user-defined type, you need to walk down the chain until you rach a built-in type:

> original_type, if and only if no @p4runtime_translation annotation is present. It specifies the underlying built-in P4 type for the user-defined type. If the underlying type used in the P4 type declaration is itself a user-defined type, original_type is obtained by “walking” the chain of type declarations recursively until a built-in type (e.g bit<W>) is found.

That means that the `new_types` map entries in the P4TypeInfo message must always map to a built-in type, not another user-defined type.

**Comment by antoninbas** [comment]
And regarding what should appear in the `new_types` section, I believe it should only be types that are used somewhere **in the P4Info**.

**Comment by hesingh** [comment]
Here is an example to make sure what should be the type_name for a MatchField.

Is type_name PortIdUint_t or PortId_t?

```
typedef bit<32>    PortIdUint_t;
type PortIdUint_t  PortId_t;
```

**Comment by antoninbas** [comment]
I'm also in favor of applying the spec literally as far as the `@p4runtime_translation` annotation is concerned. In the event of a "chain" of user-defined types `T1` -> `T2` -> `bit<W>`.
If an expression of type `T1` is used (e.g. as a match field), and `T1` has no `@p4runtime_translation` annotations, then the expressions is not considered "translated", even if `T2` has a `@p4runtime_translation` annotation.
So basically I don't think the annotation should propagate "up the chain".

However, if `T1` is a typedef to `T2`, then the expression is translated, and that's a major difference between typedef's and user-defined types.

In general I cannot think of a good use case for chains of user-defined types.

**Comment by antoninbas** [comment]
> Here is an example to make sure what should be the type_name for a MatchField.
> 
> Is type_name PortIdUint_t or PortId_t?
> 
> ```
> typedef bit<32>    PortIdUint_t;
> type PortIdUint_t  PortId_t;
> ```

`PortIdUint_t` is a typedef, it's not an actual type and it can therefore never appear in the P4Info. `type_name` is strictly for user-defined types.

**Comment by hesingh** [comment]
Right, but then `PortIdUint_t` is reaching a built-in type.  So the caveat is that if a built-in type is reached via a typedef, use the parent of the typedef which is `PortId_t`?

**Comment by jafingerhut** [comment]
@hesingh If you want to ask questions with a term like "parent" here, I think you need to give an example of which way the "parent-child" relationship goes, in your thinking.  You might be thinking the opposite direction that we are.

**Comment by antoninbas** [comment]
I don't think there is any "caveat" here. Either an expression is of type `PortId_t` or it's not. There is no such thing as an expression of type `PortIdUint_t` because it's not a type, it's an alias.
If an expression is of type `PortId_t`, the underlying **type** is `bit<32>`, not `PortIdUint_t`, which is not a type and has no meaning for P4Runtime.

**Comment by hesingh** [comment]
Understood, thanks, Antonin.   In new_types, PortId_t has been used as the code.  I will look into type_name changes in the PR to comply with today's notes.

**Comment by jafingerhut** [comment]
Antonin, thanks for the review comments.

I have not yet thought of a practical use case where it is beneficial to "stack" `type` definitions in P4 programs, either.  I only bring them up because I think it is a good idea that either:

(a) the compiler rejects them, or
(b) we say explicitly that P4Runtime does not support P4_16 programs with such type definitions, or
(c) we define precisely what the behavior is when a program contains such types.

I have been going down road (c) so far, to see where it leads and what makes sense there.

So for the following code example:
```
@p4runtime_translation("foo", 32)
type bit<9> T1;

type T1 T2;
```
If there is a bit-constrained value (i.e. table search key field, action parameter, etc.) that has type T2, you are saying that it should have `type_name: "T2"` in the P4Info file, yes?

Would you say its `bitwidth`should be 9 or 32 in the P4Info file?  If 32, then we have to pay attention to the p4runtime_translation on T1 somehow.  If 9, then that sounds more consistent with your statements above, and I can't think of any objections, preferably with a mention in some docs somewhere.

Would you expect the following in the `type_info` message of that P4Info file, or something else?
```
type_info {
  new_types {
    key: "T1"
    value {
      translated_type {
        uri: "foo"
        sdn_bitwidth: 32
      }
    }
  }
  new_types {
    key: "T2"
    value {
      original_type {
        bitstring {
          bit {
            bitwidth: 9
          }
        }
      }
    }
  }
}
```

**Comment by antoninbas** [comment]
@jafingerhut That's the `type_info` I would expect and that's what's mandated by the spec, but perhaps I'm biased since I wrote that section.

**Comment by jafingerhut** [comment]
You are biased in a very educated and informed direction, I would say :-)

I have updated the README based on your comments, and I hope it matches the behavior you expect now.

Latest version is at this link: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files

If you would prefer to see the diffs, they are here, although looking at them I am not sure if it is easier to read them, or just the latest version linked above.  I fixed up some wrong P4Info message examples at the same time, which may confuse the diffs a bit: https://github.com/jafingerhut/p4-guide/commit/d38695ecce9928d59933d6ea238e173f8b1b625c

**Comment by antoninbas** [comment]
I read https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files and I think I agree with everything there at this stage.
If you are interested, I would be in favor of including this information either as a separate document in this repository ("Guidance for generating the P4Info message"?) or directly as an appendix in the spec.

**Comment by jafingerhut** [comment]
Thanks for taking the time to read it over.

I have a slight preference for making it a separate document in this repository, if that would be OK with you.  It could be linked from the spec to make it easier for people to find it.  But I am OK with either choice.

I would like to add a little bit more material there, now that we have some consensus on what is there so far.

For example, a few notes on how to generate the type_info message in general, and including that message in all of the examples.

Also some examples for digest messages and register array element contents, which can have more arbitrary types.  Simple commonly expected kinds of examples are best to start with, but for implementers, having at least one or two wacky crazy ones can help explain the desired behavior in general.

**Comment by antoninbas** [comment]
A separate document is fine, especially if it keeps growing :)

**Comment by jafingerhut** [comment]
Take an example of a value defined with type `@p4runtime_translation("my_uri", 32) type bit<9> PortId_t` and a table search key field with that type.

I think we have established that the P4Info `type_info` message should have a `new_types` sub-message that gives the uri of "my_uri" and sdn_bitwidth of 32.

I think we have also said that the table search key field also have a bitwidth field value of 32, as well as a `type_name {  name: "PortId_t" }` field.  This bitwidth field value of 32 should be redundant, since it can be determined from looking up the information about the type named "PortId_t" elsewhere in the P4Info message.

Do we want this redundancy?  Or do we want to say that the `bitwidth` field should be unset in a case where it has a `type_info` field from which the bitwidth can be calculated?

**Comment by jafingerhut** [comment]
Created PR: https://github.com/p4lang/p4runtime/pull/193

It contains a slightly extended version of this document that Antonin reviewed, mentioned in an earlier comment on this issue: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files

The main additions are more links to Github issues for any questions that I think are still outstanding, and the addition of example 4, which is right now mostly questions on how to represent serializable enum types in a P4Info message.

**Comment by jafingerhut** [comment]
Closing this issue, since corresponding PR was merged: https://github.com/p4lang/p4runtime/pull/193
